KAFKA-19390: backport to 3.9

Forest0923 opened the pull request:
This backports KAFKA-19390 to kafka v3.9.

Reviewers: Jun Rao <junrao@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

Forest0923 committed:
backport KAFKA-19390

Forest0923 commented:
@junrao Thanks for the suggestions. I've created PR to backport the fix. Could you please review it?

chia7712 reviewed (commented):
@Forest0923 thanks for this backport!

junrao reviewed (commented):
@Forest0923 : Thanks for the backporting. The code LGTM. Do you know why the build for JDK 21 failed?

Forest0923 commented:
It seems to be a problem with Gradle's cache.
https://github.com/gradle/gradle/issues/28974

chia7712 commented:
```
Gradle Test Run :storage:test > Gradle Test Executor 25 > OffsetIndexTest > forceUnmapTest() FAILED
    org.opentest4j.AssertionFailedError: Expected java.lang.NullPointerException to be thrown, but nothing was thrown.
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:152)
        at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:73)
        at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
        at app//org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3115)
        at app//org.apache.kafka.storage.internals.log.OffsetIndexTest.forceUnmapTest(OffsetIndexTest.java:228)
```
please fix the failed test

Forest0923 committed:
test: remove assertThrows when closing offsetIndex

- update to match trunk branch
- callee methods inside close() have null guards to prevent NPE

Forest0923 commented:
@chia7712 I've updated the test code to match the code in the trunk branch.

https://github.com/apache/kafka/blob/1e4b8a1a6b9ca8afd9635f90ad3116f68ee3da94/storage/src/test/java/org/apache/kafka/storage/internals/log/OffsetIndexTest.java#L222-L228

junrao commented:
@Forest0923 : Thanks for the updated PR. Are the following test failures for JDK 23 related to the PR?

 ```
org.apache.kafka.clients.consumer.internals.AbstractCoordinatorTest.testWakeupAfterSyncGroupReceivedExternalCompletion()	32 ms	1
 org.apache.kafka.controller.AclControlManagerTest.testDeleteExceedsMaxRecords()	1 min 6 sec	1
 Gradle Test Executor 89.failed to execute tests	2 hr 38 min	1
 org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
```

Also, the tests for JDK 8 timed out.

chia7712 commented:
`AclControlManagerTest.testDeleteExceedsMaxRecords` and `CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated` already has jira, and other failed tests pass on my local. 

the timeout test is `testConsumerRebalanceListenerAssignmentOnPartitionsRevoked`, but I can't reproduce it on my local.

I trigger a re-run

junrao commented:
Filed https://issues.apache.org/jira/browse/INFRA-27278 to track the gradle cache issue for JDK21.

Forest0923 commented:
Thanks for the support.

A new failure regarding `Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.SmokeTestDriverIntegrationTest."shouldWorkWithRebalance(boolean, boolean).stateUpdaterEnabled=true, processingThreadsEnabled=true` has also been reported in Jira.

omkreddy commented:
I’ll assume we’ll be backporting the fixes to the 4.1 and 4.0 branches as well.

chia7712 commented:
> I’ll assume we’ll be backporting the fixes to the 4.1 and 4.0 branches as well.

I will cherry-pick them to 4.1 and 4.0

chia7712 commented:
@Forest0923 Could you please open a PR for 4.0? It has some conflicts.
